File: backend/venue_review.py
"""
What the moderator said — so a declined partner can act on it.

DROP-IN: paste these into `shotright/api.py`, add four fields to `Venue`, and
add the `Venue Fix Item` child table described at the bottom.

Why this exists
---------------
A decline is the only moment in this product where the partner is not the one
driving. They filled in a long form, submitted a venue, waited days, and were
told no by somebody they have never spoken to.

Right now the portal can tell them THAT and nothing else, because the decision
is a workflow state and nothing more. There is no field for a moderator to
write into, so there is no reason to show. The new review screen therefore has
to say, in as many words, "no reason was recorded" — which is honest, and is a
bad thing for a business owner to read about their own livelihood.

What we are NOT doing while we wait
-----------------------------------
Not filling the gap with a generic line. "Your venue didn't meet our
guidelines" would remove the awkwardness and replace it with something worse: a
partner acts on it, changes the wrong thing, resubmits, and is declined again —
and now they have also learned that the portal's explanations are not worth
reading. An admitted absence is recoverable. An invented reason is not.

What the screen already does without you
----------------------------------------
It derives what it can see for itself — no map pin, no photos, nothing on the
menu, no description — and shows it under its own heading, explicitly labelled
as *our* observation and not the reviewer's. Those are the common decline
reasons and they are checkable from data the portal already holds. It is a
decent consolation prize. It is not a substitute for the moderator's actual
sentence, because the reviewer may have declined the venue for something no
amount of client-side checking can see.

The one thing that will silently break this
-------------------------------------------
`frappe.call()` DROPS kwargs that are not in the declared signature — at HTTP
200, with no error. Three separate bugs on this project came from exactly that.
Keep `venue_name`, `item` and `done` as written, or tell the frontend what they
actually are.

Two things worth doing in the Desk, not here
--------------------------------------------
1. Make `review_notes` REQUIRED on the decline transition. A decline with an
   empty note is the thing this whole file exists to prevent, and a workflow
   that permits it will produce it.
2. Show the vendor's own words back to the moderator. Most declines are a
   misunderstanding in one direction or the other.
"""

import frappe
from frappe import _


def _vendor():
    """The Vendor Profile for the current session, or 403."""
    vendor = frappe.db.get_value("Vendor Profile", {"user": frappe.session.user}, "name")
    if not vendor:
        frappe.throw(_("No vendor profile for this account."), frappe.PermissionError)
    return vendor


def _owned(venue_name):
    """The Venue if it belongs to the caller, else 404 (not 403 — see below).

    404 rather than 403 on someone else's venue: a 403 confirms the venue
    exists, which turns this endpoint into a directory of every business on the
    platform for anyone willing to iterate names.
    """
    venue = frappe.db.get_value("Venue", venue_name, ["name", "vendor"], as_dict=True)
    if not venue or venue.vendor != _vendor():
        frappe.throw(_("Venue not found."), frappe.DoesNotExistError)
    return venue


@frappe.whitelist()
def get_venue_review(venue_name):
    """The current decision on a venue, and what the reviewer asked for.

    Returns:

        {
          "state":          "Declined",
          "notes":          "<the moderator's own words, plain text>",
          "reviewed_by":    "nandi@shotright.co.za",
          "reviewed_by_name": "Nandi M.",
          "reviewed_on":    "2026-07-25 14:02:11",
          "fix_items": [ {"name": "abc123", "label": "Drop the pin on the venue",
                          "done": 0}, ... ]
        }

    `notes` is shown to the partner VERBATIM, in a quote block, attributed to
    the reviewer. Write it as a message to a business owner rather than as an
    internal moderation note — it is not summarised, softened or reformatted on
    the way through, deliberately, because the partner came to the page to read
    exactly what was said.

    PLAIN TEXT, not HTML. It renders with `white-space: pre-line`, so newlines
    survive and nothing else is interpreted. Sending HTML here would either be
    escaped and shown as tags, or become an injection route into the partner's
    browser — neither is what anyone wants.

    Returning `None` is legitimate and handled: the screen says no reason was
    recorded rather than inventing one.
    """
    venue = _owned(venue_name)
    doc = frappe.get_doc("Venue", venue.name)

    if not (doc.get("review_notes") or doc.get("reviewed_on")):
        return None

    reviewer = doc.get("reviewed_by")

    return {
        "state": doc.get("workflow_state"),
        "notes": doc.get("review_notes") or "",
        "reviewed_by": reviewer,
        # The person, not the login. "Nandi M." reads as a human being having
        # looked at your restaurant; "nandi@shotright.co.za" reads as a system.
        "reviewed_by_name": frappe.db.get_value("User", reviewer, "full_name") if reviewer else None,
        "reviewed_on": doc.get("reviewed_on"),
        "fix_items": [
            {"name": row.name, "label": row.label, "done": int(row.done or 0)}
            for row in (doc.get("fix_items") or [])
        ],
    }


@frappe.whitelist()
def set_review_fix_item(venue_name, item, done):
    """Tick one fix item off.

    This is the PARTNER'S working memory, not a report to the reviewer, and the
    portal says so on screen. Do not wire it to a notification, and do not let
    it affect the workflow state — a partner who believes ticking a box has told
    somebody they fixed it will sit waiting for a reply that is not coming.

    Falls back to the partner's own browser when this method is absent, so the
    checklist works either way; deploying this just makes it follow them to
    another device.
    """
    venue = _owned(venue_name)
    row = frappe.db.get_value(
        "Venue Fix Item", {"name": item, "parent": venue.name}, "name"
    )
    if not row:
        frappe.throw(_("That item is not on this venue."), frappe.DoesNotExistError)

    frappe.db.set_value("Venue Fix Item", row, "done", 1 if int(done or 0) else 0)
    frappe.db.commit()
    return {"ok": True}


# ---------------------------------------------------------------------------
# DOCTYPES
# ---------------------------------------------------------------------------
#
# On "Venue", add:
#
#   review_notes   Small Text   What the moderator wants to say to the partner.
#                               PLAIN TEXT. ⚠️ Make this REQUIRED on the decline
#                               transition — see the note at the top.
#   reviewed_by    Link → User  Set by the workflow, read-only.
#   reviewed_on    Datetime     Set by the workflow, read-only.
#   fix_items      Table → Venue Fix Item
#
# Child table: "Venue Fix Item"  (istable = 1)
#
#   label   Data    One thing to change, phrased as an instruction the partner
#                   can act on: "Drop the pin on the venue", not "location".
#   done    Check   Ticked by the PARTNER, not the moderator.
#
# PERMISSIONS
#
#   The Vendor role needs no direct permission on any of this — both methods
#   resolve ownership from the session. `review_notes` must NOT be vendor-
#   writable: it is the reviewer's statement, and a partner able to edit it
#   would make the record useless to both of them.
#
# THE DECISION WE NEED FROM YOU
#
#   Does resubmitting EDIT THE VENUE, or restore a DRAFT?
#
#   We think editing the Venue is right: the review history hangs off it, the
#   partner keeps one thing with one identity, and "edit and resubmit" is one
#   action rather than a copy that has to be reconciled later. The portal is
#   built that way — the button on the review screen goes to the venue's own
#   edit form.
#
#   If you would rather it became a Draft again, `save_venue_draft` needs a
#   `venue` link field so the two can be reconnected, and the review screen's
#   button changes. Tell us before you build it either way, because those are
#   different products and only one of them matches what is shipped.
#
# AND ONE SMALL ONE
#
#   `reviewed_on` is rendered as a date, not a timestamp. If your workflow
#   stamps it on every save rather than only on a decision, it will read as
#   though someone reviewed the venue again today. Stamp it on the transition.

File: src/services/venueReview.js
import { callGet, call, USE_MOCKS, isMethodMissing } from './api'
import { mockBackend } from './mockBackend'
import { withFallback } from './vendor'

/**
 * What the moderator said, and what to do about it.
 *
 * A decline is the only moment in this product where the partner is not the one
 * driving. They submitted a venue, waited, and were told no. Everything on the
 * screen this feeds exists to answer one question — *why, and what do I change?*
 *
 * THE HONEST FAILURE MODE MATTERS MORE THAN THE HAPPY ONE. Where a reason
 * genuinely wasn't recorded we say so, rather than softening it with a generic
 * "your venue didn't meet our guidelines" — a made-up reason is worse than an
 * admitted absence, because a partner acts on it, changes the wrong thing, and
 * is declined again.
 *
 * ⚠️ CORRECTED 28 JUL — AND THE CORRECTION IS THE INTERESTING PART.
 *
 * This file was written believing the bench had no field for a moderator to
 * write into. It has had three all along: `review_notes`, `reviewed_by` and
 * `reviewed_on` are on the `Venue` doctype, and `get_vendor_dashboard` has been
 * returning them to this portal the entire time — put there deliberately so a
 * declined venue could render its reasons without a second round trip.
 *
 * So the portal was printing "No reason was recorded" on top of a reason it had
 * already been handed. The capability check was sound and the answer it gave
 * was true — `get_venue_review` really isn't deployed — but it was the answer
 * to a question that didn't need asking. A missing ENDPOINT was reported to the
 * partner as missing DATA.
 *
 * The lesson generalises past this screen: check what you already hold before
 * you ask whether you can get it. The read order below does that.
 *
 *   1. `get_venue_review` — if ever deployed, the richest source, and the only
 *      one that can carry `fix_items`.
 *   2. THE VENUE RECORD WE ALREADY HAVE. This is the one that works today.
 *   3. The dashboard row — for the case where `get_venue_detail` doesn't pass
 *      the review fields through but `get_vendor_dashboard` does.
 *
 * First source actually carrying the field wins. Nothing needs to be deployed
 * for a partner to read why they were declined.
 */
export const REVIEW_READ_METHOD = 'shotright.api.get_venue_review'
export const FIX_ITEM_METHOD = 'shotright.api.set_review_fix_item'

/**
 * Where "Contact support" goes.
 *
 * Deliberately NOT defaulted to a guessed address. Every invented string on
 * this project — `vendor_name`, `Rejected`, five menu method names — became a
 * bug that reached a partner, and a support address nobody has confirmed is
 * the worst of them: a message to it does not bounce loudly, it just never
 * gets read. Unset means the button is not shown and the screen says so.
 */
export const SUPPORT_EMAIL = import.meta.env.VITE_SUPPORT_EMAIL || ''

const FIX_STORE = 'shotright.fixItems'

const readStore = () => {
  try {
    return JSON.parse(localStorage.getItem(FIX_STORE) || '{}')
  } catch {
    return {}
  }
}

/**
 * Does this payload carry review fields at all?
 *
 * The distinction that matters is between a field that is PRESENT AND EMPTY —
 * nobody wrote a note — and a field that isn't in the payload, meaning whatever
 * we read simply doesn't pass it through. To the partner both render as
 * silence; to us one is a process gap and the other is our own plumbing, and we
 * must not blame the reviewer for the second.
 *
 * Frappe sends an empty Small Text as `""` or `null`, so presence of the KEY is
 * the test — never truthiness of the value.
 */
const carriesReview = (raw) =>
  Boolean(raw) &&
  ['notes', 'review_notes', 'reviewed_on', 'reviewed_by'].some((key) => key in raw)

/**
 * The review, normalised.
 *
 * `available` now means "we could read the review fields", NOT "an endpoint
 * answered" — those came apart the moment we learned the fields ride along on
 * the venue record. `source` names where it came from, so a bug report can say
 * which of three paths produced the screen someone is looking at.
 */
const normalise = (raw, source) => {
  const notes = String(raw?.notes ?? raw?.review_notes ?? '').trim()
  return {
    available: source !== 'none',
    source,
    notes,
    reviewedBy: raw?.reviewed_by_name || raw?.reviewed_by || '',
    reviewedOn: raw?.reviewed_on || '',
    state: raw?.state || raw?.workflow_state || '',
    fixItems: (raw?.fix_items || []).map((item, index) => ({
      key: item.name || item.key || `item-${index}`,
      label: item.label || item.description || String(item),
      done: Boolean(item.done),
    })),
  }
}

/**
 * `reviewed_on` used to fall back to `modified`. It must not.
 *
 * The bench stamps `reviewed_on` on the workflow TRANSITION, so it means "the
 * day someone decided". `modified` means "the day the row last changed" — which
 * a moderator opening the doc in October will move, and the screen renders it
 * as "· 12 October" under the reviewer's name. A partner declined in July would
 * read that as having been judged again. A missing date is better than a wrong
 * one, so an absent `reviewed_on` now shows no date at all.
 */

/**
 * @param venueId  the docname
 * @param venue    the venue record the calling screen already holds, if any.
 *                 Passing it is what makes this work with nothing deployed.
 */
export const getVenueReview = async (venueId, venue = null) => {
  if (USE_MOCKS) return normalise(await mockBackend.getVenueReview(venueId), 'endpoint')

  // 1. The dedicated endpoint, if it is ever deployed. Only source of fix_items.
  const fromEndpoint = await withFallback(
    REVIEW_READ_METHOD,
    async () => await callGet(REVIEW_READ_METHOD, { venue_name: venueId }),
    async () => null,
  ).catch(() => null)
  if (carriesReview(fromEndpoint)) return normalise(fromEndpoint, 'endpoint')

  // 2. The record in hand. No round trip, and it is what ships today.
  if (carriesReview(venue)) return normalise(venue, 'venue')

  // 3. The dashboard row — `get_vendor_dashboard` is confirmed to carry the
  //    review fields; `get_venue_detail` is not, and the screen reads detail.
  try {
    const dash = await call('shotright.api.get_vendor_dashboard')
    const row = (dash?.venues || []).find(
      (v) => v?.name === venueId || v?.venue_name === venueId,
    )
    if (carriesReview(row)) return normalise(row, 'dashboard')
  } catch {
    // A failed dashboard read must not take down a screen that already has the
    // venue, the derived gaps and the way to reach a human.
  }

  return normalise(null, 'none')
}

/**
 * Tick a fix item off.
 *
 * Falls back to this browser, like drafts do — the checklist is the partner's
 * working notes, and losing them on another device is a smaller harm than not
 * being able to keep track at all. It is not, and must not be presented as,
 * something the reviewer can see.
 */
export const setFixItemDone = async (venueId, key, done) => {
  if (!USE_MOCKS) {
    try {
      await call(FIX_ITEM_METHOD, { venue_name: venueId, item: key, done: done ? 1 : 0 })
      return { synced: true }
    } catch (err) {
      if (!isMethodMissing(err, FIX_ITEM_METHOD)) throw err
    }
  }
  const store = readStore()
  store[venueId] = { ...(store[venueId] || {}), [key]: done }
  localStorage.setItem(FIX_STORE, JSON.stringify(store))
  return { synced: false }
}

export const localFixState = (venueId) => readStore()[venueId] || {}

/* --------------------------------------------------------------- our own */

/**
 * Gaps in the listing that we can see without being told.
 *
 * These are NOT the reviewer's reasons and the screen must never present them
 * as such — that is why they come from a separate function, render under their
 * own heading, and are worded as observations rather than instructions from
 * anybody. But they are the four or five things a listing is actually declined
 * for, they are checkable from data already on the page, and when the reviewer
 * left no note they are the only concrete thing we can offer.
 *
 * Ordered by how much each one costs the partner. A venue with no coordinates
 * cannot be found by anyone at all; a missing dress code is cosmetic.
 */
export function deriveGaps(venue, { photos = [], menu = [] } = {}) {
  const gaps = []
  const text = (value) => String(value || '').replace(/<[^>]*>/g, '').trim()

  if (!Number.isFinite(venue?.latitude) || !Number.isFinite(venue?.longitude)) {
    gaps.push({
      key: 'location',
      label: 'No location is set',
      detail:
        'Customers find venues by searching near themselves, so a venue with no pin never appears — whatever else is on it.',
      to: 'edit',
    })
  }

  if (!(venue?.moods || []).length) {
    gaps.push({
      key: 'moods',
      label: 'No moods are set',
      detail: 'Sho’t Right searches by mood. A venue with none cannot be matched to anything.',
      to: 'edit',
    })
  }

  if (!photos.length) {
    gaps.push({
      key: 'photos',
      label: 'No photos',
      detail:
        'A listing with no picture asks someone to pick where to spend their evening on a name alone.',
      to: 'edit',
    })
  }

  const itemCount = menu.reduce((n, heading) => n + (heading.items?.length || 0), 0)
  if (!itemCount) {
    gaps.push({
      key: 'menu',
      label: 'Nothing on the menu',
      detail: 'Even a handful of items and prices gives someone a reason to choose you.',
      to: 'menu',
    })
  }

  if (!text(venue?.atmosphere_desc) && !text(venue?.summary)) {
    gaps.push({
      key: 'description',
      label: 'No description',
      detail: 'A line or two about what a night here is like.',
      to: 'edit',
    })
  }

  return gaps
}

/**
 * A prefilled support email.
 *
 * Carries the venue id and the decline date because the first two replies of
 * any support thread are otherwise "which venue?" and "when?". The partner's
 * own message goes at the top, where they will type it, not appended under a
 * block of metadata they have to scroll past.
 */
export function supportMailto({ venueName, venueId, reviewedOn, message = '' }) {
  if (!SUPPORT_EMAIL) return null
  const body = [
    message,
    '',
    '---',
    `Venue: ${venueName || venueId}`,
    `Reference: ${venueId}`,
    reviewedOn ? `Declined on: ${new Date(reviewedOn).toLocaleDateString('en-ZA')}` : null,
  ]
    .filter((line) => line !== null)
    .join('\n')

  return (
    `mailto:${SUPPORT_EMAIL}` +
    `?subject=${encodeURIComponent(`Declined venue: ${venueName || venueId}`)}` +
    `&body=${encodeURIComponent(body)}`
  )
}
